src/vaadin-grid-drag-and-drop-mixin.html

```diff
@@ -0,0 +1,257 @@
+
+  (function() {
+    const DropMode = {
+      BETWEEN: 'between',
+      ON_TOP: 'on-top',
+      ON_TOP_OR_BETWEEN: 'on-top-or-between',
+      ON_GRID: 'on-grid'
+    };
+
+    const DropLocation = {
+      ON_TOP: 'on-top',
+      ABOVE: 'above',
+      BELOW: 'below',
+      EMPTY: 'empty'
+    };
+
+    /**
+     * @polymerMixin
+     */
+    Vaadin.Grid.DragAndDropMixin = superClass => class DragAndDropMixin extends superClass {
+
+      static get properties() {
+        return {
+          /**
+           * Defines the locations within the Grid row where an element can be dropped.
+           *
+           * Possible values are:
+           * `between`: The drop event can happen between Grid rows.
+           * `on-top`: The drop event can happen on top of Grid rows.
+           * `on-top-or-between`: The drop event can happen either on top of or between Grid rows.
+           * `on-grid`: The drop event will not happen on any specific row, it will show the drop target outline around the whole grid.
+           */
+          dropMode: String,
+
+          /**
+           * Marks the grid's rows to be available for dragging.
+           */
+          rowsDraggable: Boolean
+
+        };
+      }
+
+      static get observers() {
+        return [
+          '_rowsDraggableChanged(rowsDraggable)'
+        ];
+      }
+
+      ready() {
+        super.ready();
+        this.$.table.addEventListener('dragstart', this._onDragStart.bind(this));
+        this.$.table.addEventListener('dragend', this._onDragEnd.bind(this));
+        this.$.table.addEventListener('dragover', this._onDragOver.bind(this));
+        this.$.table.addEventListener('dragleave', this._onDragLeave.bind(this));
+        this.$.table.addEventListener('drop', this._onDrop.bind(this));
+      }
+
+      _rowsDraggableChanged() {
+        this._assignModels();
+      }
+
+      _onDragStart(e) {
+
+        e.stopPropagation();
+        this._toggleAttribute('dragging-rows', true, this);
+
+        const item = this.getEventContext(e).item || this.__getRowModel(e.target).item;
+
+        const row = Array.from(this.$.items.children).filter(row => row._item === item)[0];
+        if (!window.ShadyDOM) {
+          // Safari doesn't get proper drag images from transformed
+          // elements so we need to switch to top temporarily
+          const transform = row.style.transform;
+          row.style.top = /translateY\((.*)\)/.exec(transform)[1];
+          row.style.transform = 'none';
+
+          const rect = row.getBoundingClientRect();
+          e.dataTransfer.setDragImage(row, e.clientX - rect.left, e.clientY - rect.top);
+
+          requestAnimationFrame(() => {
+            row.style.top = '';
+            row.style.transform = transform;
+          });
+        }
+
+        const items = this._isSelected(item) ? this.selectedItems.slice(0) : [item];
+        e.dataTransfer.setData('text', this.formatDragData(items));
```

Comment:
One funny observation: Dragging on IE11 copies the drag data to clipboard. Just pointing out a weird browser implementation.